Analysis/bin/ShuffleMergeSpectral.cxx

```diff
@@ -0,0 +1,775 @@
+                  << " scale factor: " << dis_scale << "\n";
+        ttype_prob.at(type)->Scale(dis_scale/ttype_prob.at(type)->GetBinContent(x,y));
+      }
+    }
+
+    const std::map<TauType, Double_t> GetTauTypeEntriesFinal() const
+    {
+      std::map<TauType, Double_t> EntriesFinal;
+      for (TauType type: ttypes) {
+        std::shared_ptr<TH2D> entries_count((TH2D*)ttype_entries.at(type)->Clone());
+        for(int i_x = 1; i_x<=ttype_prob.at(type)->GetNbinsX(); i_x++)
+          for(int i_y = 1; i_y<=ttype_prob.at(type)->GetNbinsY(); i_y++)
+            entries_count->SetBinContent(i_x,i_y,
+              entries_count->GetBinContent(i_x,i_y)*ttype_prob.at(type)->GetBinContent(i_x,i_y)
+            );
+        EntriesFinal[type] = entries_count->Integral();
+      }
+      return EntriesFinal;
+    }
+
+    const double GetProbability(const TauType& type, const double& pt, const double& eta) const
+    {
+      return ttype_prob.at(type)->GetBinContent(
+             ttype_prob.at(type)->GetXaxis()->FindBin(eta),
+             ttype_prob.at(type)->GetYaxis()->FindBin(pt));
+    }
+
+    void SaveHists(const std::string& output)
+    {
+      if(!boost::filesystem::exists(output)) boost::filesystem::create_directory(output);
+      TFile file = TFile((output+"/"+groupname+".root").c_str(),"RECREATE");
+      for (TauType type: ttypes)
+        ttype_prob[type]->Write((analysis::ToString(type)+"_prob").c_str());
+      file.Close();
+      TFile file_n = TFile((output+"/"+groupname+"_n.root").c_str(),"RECREATE");
+      for (TauType type: ttypes)
+        ttype_entries[type]->Write((analysis::ToString(type)+"_entries").c_str());
+      file_n.Close();
+    }
+
+    const size_t GetEntries() { return n_entries; }
+    const std::string GetGroupName() { return groupname; }
+
+private:
+
+  std::shared_ptr<TH2D> SetTargetHist_test(const double scale,
+                                           const std::vector<double>& pt_bins,
+                                           const std::vector<double>& eta_bins)
+  {
+    std::shared_ptr<TH2D> tartget_hist = std::make_shared<TH2D>("tartget","tartget",
+                                          eta_bins.size()-1,&eta_bins[0],
+                                          pt_bins.size()-2,&pt_bins[0]);
+    for(Int_t i_pt = 1; i_pt <= tartget_hist->GetNbinsY(); i_pt++)
+      for(Int_t i_eta = 1; i_eta <= tartget_hist->GetNbinsX(); i_eta++)
+        tartget_hist->SetBinContent(i_eta,i_pt,1.0);
+    tartget_hist->Scale(scale/tartget_hist->Integral());
+    return tartget_hist;
+  }
+
+  static std::pair<std::vector<int>,std::vector<int>> GetBinMap(const std::shared_ptr<TH2D>& newbins,
+                                                                const std::shared_ptr<TH2D>& oldbins)
+  {
+    // check over x axis
+    std::vector<int> x_map;
+    for(Int_t new_x=0; new_x<=newbins->GetNbinsX(); new_x++){
+      float up_x = (float)newbins->GetXaxis()->GetBinUpEdge(new_x);
```

Comment:
> "why conversion to float is necessary?"
Sometimes `->GetBinUpEdge` method when for instance having edge at 1.5 it returns 1.5(00).....1 when using Double_t. Conversion to float was a dirty step to avoid it.